### PR TITLE
omnistaging

### PR DESCRIPTION
**Pull request by mattjj**
### Motivation

JAX transformations like `jit` and `pmap` stage out computations to XLA. That is, we apply them to functions comprising multiple `jax.numpy` operations so that rather being executed one at a time from Python the operations are all part of one end-to-end optimized XLA computation.

But exactly which operations get staged out? Here’s an example:

```python
import jax.numpy as jnp
from jax import lax

@jit
def select_tril(x):
  mask = jnp.arange(x.shape[0])[:, None] > jnp.arange(x.shape[1])
  return lax.select(mask, x, jnp.zeros_like(x))  # lax.select is like jnp.where

x = np.arange(12).reshape((3, 4))
select_tril(x)
```

```
ENTRY jit_select_tril.8 {
  constant.3 = pred[] constant(false)
  constant.1 = pred[3,4]{1,0} constant({...})
  parameter.2 = s32[3,4]{1,0} parameter(0)
  constant.4 = s32[] constant(0)
  broadcast.5 = s32[3,4]{1,0} broadcast(constant.4), dimensions={}
  select.6 = s32[3,4]{1,0} select(constant.1, parameter.2, broadcast.5)
  ROOT tuple.7 = (s32[3,4]{1,0}) tuple(select.6)
}
```

The `select` operation is staged out, but the operations for constructing the constant `mask` are not. Rather than being staged out, the operations that construct `mask` are executed op-by-op at Python tracing time, and XLA only sees a compile time constant `constant.1` representing the value of `mask`. That’s unfortunate, because if we had staged out the operations for constructing `mask`, XLA could have fused them into the `select` and avoided materializing the result at all. **As a result we end up wasting memory with a large constant, wasting time dispatching multiple un-fused op-by-op XLA computations, and potentially even fragmenting memory.**

(The `broadcast` that corresponds to the construction of the zeros array for `jnp.zeros_like(x)` is staged out because JAX is lazy about very simple expressions from #1668.)

The reason the creation of `mask` is not staged out is that `jit` operates based on data dependence. That is, `jit` stages out only those operations in a function that have a data dependence on an augment. Control flow primitives and `pmap` behave similarly. In the case of `select_tril`, the operations to construct the constant `mask` do not have a data dependence on the argument `x`, so they are not staged out; only the `lax.select` call has a data dependence.

The omnistaging change is about enabling `jit`, `pmap`, and control flow primitives to stage out more computation to XLA. As the name implies, it’s about staging out as much as possible! More precisely, with omnistaging all `jax.numpy` calls in the dynamic context of a `jit`-transformed function are staged out to XLA.

After omnistaging, the computation XLA sees for `select_tril` is

```
ENTRY jit_select_tril.16 {
  constant.4 = pred[] constant(false)
  iota.1 = s32[3]{0} iota(), iota_dimension=0
  broadcast.5 = s32[3,1]{1,0} broadcast(iota.1), dimensions={0}
  reshape.7 = s32[3]{0} reshape(broadcast.5)
  broadcast.8 = s32[3,4]{1,0} broadcast(reshape.7), dimensions={0}
  iota.2 = s32[4]{0} iota(), iota_dimension=0
  broadcast.6 = s32[1,4]{1,0} broadcast(iota.2), dimensions={1}
  reshape.9 = s32[4]{0} reshape(broadcast.6)
  broadcast.10 = s32[3,4]{1,0} broadcast(reshape.9), dimensions={1}
  compare.11 = pred[3,4]{1,0} compare(broadcast.8, broadcast.10), direction=GT
  parameter.3 = s32[3,4]{1,0} parameter(0)
  constant.12 = s32[] constant(0)
  broadcast.13 = s32[3,4]{1,0} broadcast(constant.12), dimensions={}
  select.14 = s32[3,4]{1,0} select(compare.11, parameter.3, broadcast.13)
  ROOT tuple.15 = (s32[3,4]{1,0}) tuple(select.14)
}
```

In addition to improving JAX’s memory performance, omnistaging enables a host of other improvements and simplifications throughout JAX. For example, it allows `lax.cond` to accept thunks, so that `lax.cond(x > 0, lambda: f(y), lambda: g(z))` will only evaluate one of `f(y)` and `g(z)`. Another example is that it lets us remove the lazy sublanguage of #1668.

### What's the catch? Common issues and fixes

TODO

### Enable with a flag!

This PR sets up parallel implementations of the old core and the new omnistaging one. The old one is still used by default; omnistaging is enabled by setting the `JAX_OMNISTAGING` environment variable, by setting the boolean flag `jax_omnistaging`, or by using `from jax.config import config; config.enable_omnistaging()`.

### Main implementation idea

The semantics of transformations in JAX are based on a stack of interpreters. When you call a transformed function, we think of pushing its transformations onto [a global stack](https://github.com/google/jax/blob/c38bc36803221728a05e76d962afd64c3f188c42/jax/core.py#L588). Each stack frame represents a layer of interpretation. When we [bind a primitive](https://github.com/google/jax/blob/c38bc36803221728a05e76d962afd64c3f188c42/jax/core.py#L270), we interpret it using the transformation on the top of the stack (calling the corresponding `Trace`'s `process_primitive`). If that transformation rule itself binds primitives, those binds are interpreted using the second stack frame. And so on down the stack. If the transformation rule for the interpreter on the bottom of the stack binds primitives, those binds get [interpreted with an implicit evaluation interpreter](https://github.com/google/jax/blob/c38bc36803221728a05e76d962afd64c3f188c42/jax/core.py#L274-L275). That implicit evaluation interpreter is special because we know that it will bind no primitives and that there are no more transformations that can come. It's the exit.

Those are the semantics, but there's a kind of an optimization: transformations in the stack are [only applied](https://github.com/google/jax/blob/c38bc36803221728a05e76d962afd64c3f188c42/jax/core.py#L273) if the [input arguments to a primitive bind are boxed in a `Tracer` of the corresponding transformation](https://github.com/google/jax/blob/c38bc36803221728a05e76d962afd64c3f188c42/jax/core.py#L704). (If the inputs are boxed with different Tracers, the one corresponding to the highest level in the stack wins; that's the purpose of keeping the stack around, and having `level`s!) That lets us skip irrelevant transformations for primitive binds whose inputs haven't been "infected" with the transformation. Because we've always treated `jit` the same way, it also results in `jit`'s data-dependence behavior, where the only operations that are staged out are those boxed in the appropriate `Tracer`s because they have a data dependence on an argument to the `jit`-transformed function. As we've seen, this data-dependent-staging behavior for `jit` can be undesirable.

The solution idea here is to make `jit` (and `pmap`) act differently, by making the implicit bottom of the stack less implicit. That is, we instantiate the special exit-at-the-bottom-of-the-stack interpreter frame explicitly. We [make the eval interpreter an explicit interpreter](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/core.py#L571-L580), and when the trace stack is initialized [we put an eval interpreter at the base](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/core.py#L1424-L1425). The point of making it explicit is that now we can swap it out: when we [trace a `jit`ted function](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/interpreters/xla.py#L592), we [swap out](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/interpreters/partial_eval.py#L1064) the [base of the interpretation stack](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/core.py#L1530). That way, when the interpretation of a primitive bind bottoms out while we're tracing a `jit`ted function, instead of evaluating the primitive application we'll stage it out to a jaxpr. Because [we always hit the bottom of the stack](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/core.py#L1545-L1547), [regardless](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/core.py#L719-L722) of whether arguments to the primitive bind are boxed in a particular Tracer, we end up staging out all primitive binds. Hence "omnistaging"!

(There's one more twist, but it's not really fundamental. The `dynamic` attribute on the trace stack is one additional mechanism to handle control flow primitives. For those, we want to insert a "special exit-the-system staging frame" at an arbitrary point in the trace stack, not just at the bottom. So they [trace functions to jaxprs](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/lax/lax_control_flow.py#L2423) by [pushing a `dynamic` trace to the top of the stack](https://github.com/google/jax/blob/8baa5daa3d04925dd3f437f516c00f0fcd9162a9/jax/interpreters/partial_eval.py#L1035) rather than swapping out the base. That in effect lets us point at a function and say "regardless of whatever transformations are already going on in the system, and regardless of what transformations this function will itself push onto the stack when I run it, give me a jaxpr for this function on these avals!")

For comparison, the current implementation sets up `jit` (and `pmap`) traces in [a special downward-growing stack](https://github.com/google/jax/blob/c38bc36803221728a05e76d962afd64c3f188c42/jax/core.py#L593), which in some ways also is meant always to sit under the regular transformation stack. But it's a whole stack, rather than just a single base frame, because it has to sort out data dependence issues (e.g. in the case of multi-input primitives) and decide for any given primitive bind which `jit` traces it should be staged into. Since we're not relying on data dependence anymore, we don't need a whole stack!

### CODE REVIEW

**Comment by shoyer** [comment]
@hawkinsp suggested that omnistaging might help with our slow compile times in https://github.com/google/jax/issues/3847, so I decided to give it a try.

Unfortunately omnistaging currently breaks that part of TFP used in that example. It seems to rely on eager evaluation, e.g., something like the following:
```
In [6]: jnp.concatenate([jnp.array([3]), jnp.array([3, 4])])
Out[6]: DeviceArray([3, 3, 4], dtype=int32)
```
from these lines of code:
https://github.com/tensorflow/probability/blob/7e8be7a3820aeaf023567794c02b8dbd6b7d3e84/tensorflow_probability/python/optimizer/lbfgs.py#L427-L429

With omnistaging, this stages out an abstract value, which isn't a valid input into `shape`.

**Comment by mattjj** [comment]
Yes, TFP breaks super hard with omnistaging. I've iterated with the TFP folks a good bit already. I think in the long run they'll need to fix how they call JAX code, since even now omnistaging is just turning a performance bug (potentially doing shape arithmetic on the accelerator) into a loud error, but in the short run they'll have to use a flag to disable omnistaging.

I believe @hawkinsp meant that omnsitaging would make fixing #3847 easy, and he's right! All we have to do is memoize the call on [this line](https://github.com/google/jax/blob/074583cac4a696e1378be5e54525381f12698545/jax/interpreters/partial_eval.py#L945) and we'll avoid retracing functions when we can. (All the arguments there are already hashable too!)

**Comment by shoyer** [comment]
I'm looking forward to this glorious future!

File: jax/ad_util.py [hunk 1]
```diff
@@ -27,7 +28,10 @@
 jaxval_adders[Unit] = lambda _, __: unit
 
 def add_jaxvals(x, y):
-  return add_jaxvals_p.bind(x, y)
+  if core.get_aval(x) is core.get_aval(y) is core.abstract_unit:
+    return core.unit
+  else:
+    return add_jaxvals_p.bind(x, y)
```
Comment (mattjj) [inline jax/ad_util.py:34]: I added logic to skip the bind here just because it made jaxprs cleaner to look at in some cases.

**Comment by n2cholas** [comment]
Thanks for implementing this!!

> Another example is that it lets us remove the lazy sublanguage of #1668.

To clarify, it would only allow the removal of the lazy sublanguage inside of JIT functions, right? #1668 still seems beneficial in the eager op-by-op case.

**Comment by mattjj** [comment]
It's not clear yet how much it'll make sense to remove, but once omnistaging is on by default the need for the lazy sublanguage is greatly reduced, so the complexity costs of its implementation start to look a lot bigger.
